Applications/Spire/Include/Spire/Ui/AnyInputBox.hpp

```diff
@@ -0,0 +1,183 @@
+      /**
+       * Signals that the current value is being submitted.
+       * @param submission The submitted value.
+       */
+      using SubmitSignal = Signal<void (const AnyRef& submission)>;
+
+      /**
+       * Signals that the current value was rejected as a submission.
+       * @param value The rejected value.
+       */
+      using RejectSignal = Signal<void (const AnyRef& value)>;
+
+      /**
+       * Constructs an AnyInputBox.
+       * @param input_box The input box to encapsulate.
+       * @param parent The parent widget.
+       */
+      template<typename T>
+      explicit AnyInputBox(T& input_box, QWidget* parent = nullptr);
+
+      /** Returns the current value model. */
+      const std::shared_ptr<AnyValueModel>& get_current() const;
+
+      /** Returns the last submission. */
+      const AnyRef& get_submission() const;
+
+      /** Returns <code>true</code> iff this input box is read-only. */
+      bool is_read_only() const;
+
+      /** Sets whether this input box is read-only. */
+      void set_read_only(bool read_only);
+
+      /** Connects a slot to the SubmitSignal. */
+      boost::signals2::connection connect_submit_signal(
+        const SubmitSignal::slot_type& slot) const;
+
+      /** Connects a slot to the RejectedSignal. */
+      boost::signals2::connection connect_reject_signal(
+        const RejectSignal::slot_type& slot) const;
+
+    private:
+      struct AbstractInputBox {
+        virtual ~AbstractInputBox() = default;
+        virtual const std::shared_ptr<AnyValueModel>& get_current() const = 0;
+        virtual const AnyRef& get_submission() = 0;
+        virtual bool is_read_only() const = 0;
+        virtual void set_read_only(bool read_only) = 0;
+        virtual boost::signals2::connection connect_submit_signal(
+          const SubmitSignal::slot_type& slot) const = 0;
+        virtual boost::signals2::connection connect_reject_signal(
+          const RejectSignal::slot_type& slot) const = 0;
+      };
+      template<typename T>
+      struct WrapperInputBox final : AbstractInputBox {
+        mutable SubmitSignal m_submit_signal;
+        mutable RejectSignal m_reject_signal;
+        T* m_input_box;
+        std::shared_ptr<AnyValueModel> m_current;
+        AnyRef m_submission;
+        boost::signals2::scoped_connection m_submit_connection;
+        boost::signals2::scoped_connection m_reject_connection;
+
+        WrapperInputBox(T& input_box);
+        const std::shared_ptr<AnyValueModel>& get_current() const override;
+        const AnyRef& get_submission() override;
+        bool is_read_only() const override;
+        void set_read_only(bool read_only) override;
+        boost::signals2::connection connect_submit_signal(
+          const SubmitSignal::slot_type& slot) const override;
+        boost::signals2::connection connect_reject_signal(
+          const RejectSignal::slot_type& slot) const override;
+      };
+      std::unique_ptr<AbstractInputBox> m_input_box;
```

Comment:
What's the benefit of using an abstract type here versus declaring a single concrete type?

Reply:
Here, I am using type erasure. The `AnyInputBox` is not a class template, but its constructor accepts a generic input box. We can't declare a concrete type, because the `AnyInputBox` can't hold a member variable whose type is `T`. The type erasure can hide the type of the input box, and it also can encapsulate all methods and data of the input box into the `WrapperInputBox`.

Reply:
Makes sense, thanks.